content/blog/2022-05-17-object-oriented-features/index.md

```diff
@@ -0,0 +1,210 @@
+
+We build two classes to assist with building dispatch vectors:
+
+First we need to figure out when given some class `A`, and some function `f`, which classes `f` should we call. For example, if `f` is redefined by a subclass then we want to call the `f` of that subclass. But if `f` is defined by a super class we want to call `f` of that super class. This is all handled by the `DispatchVectorClassResolver`
+
+The next thing we need is given all methods for a given class, what index do they go in in the dispatch vector. For example if `A` defines `f` and subclass `B` also defines `f` they must be at the same index. the `DispatchVectorIndexResolver` handles this.
+
+Both of these classes have been well tested.
+
+When generating the actual IR for the dispatch vectors, we tried a few different approaches. First, we converted the mangled function names into strings by representing them as an array of characters in the IR, and then tried to call the function by loading the string. We quickly realized that this would be difficult because the strings are longer often multiple characters and when we do `call [dv + index]`, we're really calling a function with the name corresponding the the first character in the mangled function name. We quickly scrapped this approach and tried to build them using static data.
+
+In Xi, static data is represented by an `IRData` class. Out of the box, `IRData` is not a traditional `IRNode` (it does not extend `IRNode`. This means its children nodes are not `IRNode`s. Instead, it has a field `long[] data`. We tried to encode the dispatch vector as the concatenation of function names seperated by the null terminator. Again, we ran into the problem that when we do `call classADVLabel` it only reads the first word of data which corresponds to the first character.
+Even worse, the interpreter tried to jump to the instruction at the address corresponding to that character instead of looking up that character in the function defn to address table and jumping there. I wrote a version of the interpreter that read until the next null terminator and then did the correct lookup. This got things to worked but it felt hacky because we could no longer write IR that jumped to an address -- it required we jumped to a function that was named as the string starting at that memory address until the next null terminator byte in memory.
+
+We explored trying to put the memory address of the function in the IRData instead of the name. That way, we would only need to read one word and the interpeter would jump there as it wanted to do. The problem with this approach is that when we convert to the IR, we have no idea about the memory layout. The memory addresses are defined by the interpeter and would be different than the memory scheme if we went all the way to assembley. Additionally, there was no way to know about the memory layout of the simulator when converting to IR because it does not initialize its memory until after the IR is created.
```

Comment:
It might be handy to clarify earlier that you are not making a native-code compiler in this project (the reader could be forgiven for assuming that).